internal/resources/openshift.go

```diff
@@ -0,0 +1,675 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package resources
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	rtclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	sambaoperatorv1alpha1 "github.com/samba-in-kubernetes/samba-operator/api/v1alpha1"
+	"github.com/samba-in-kubernetes/samba-operator/internal/conf"
+)
+
+// ClusterType reperesnts the sub-kind of k8s cluster type.
+type ClusterType string
+
+const (
+	// ClusterTypeDefault defines the default value for cluster type
+	ClusterTypeDefault = "default"
+	// ClusterTypeOpenshift defines the type-name for OpenShift clusters
+	ClusterTypeOpenshift = "openshift"
+)
+
+const (
+	openshiftFinalizer        = "samba-operator.samba.org/openshiftFinalizer"
+	serviceAccountName        = "samba"
+	sccRoleName               = "samba-scc-role"
+	sccRoleBindingName        = "samba-scc-rolebinding"
+	metricsRoleName           = "samba-metrics-role"
+	metricsRoleBindingName    = "samba-metrics-rolebinding"
+	clusterMonitoringLabelKey = "openshift.io/cluster-monitoring"
+)
+
+// OpenShiftManager is used to manage OpenShift's related resources: SCC,
+// ServiceAccount (if not exists) and enable Role and RoleBinding referencing
+// to OpenShift's 'anyuid' SCC.
+type OpenShiftManager struct {
+	client rtclient.Client
+	logger logr.Logger
+	cfg    *conf.OperatorConfig
+}
+
+// NewOpenShiftManager creates a ServiceAccountManager instance
+func NewOpenShiftManager(
+	clnt rtclient.Client,
+	log logr.Logger,
+	cfg *conf.OperatorConfig) *OpenShiftManager {
+	return &OpenShiftManager{
+		client: clnt,
+		logger: log,
+		cfg:    cfg,
+	}
+}
+
+// Process is called by the controller on reconciliation of
+// SmbCommonConfig
+func (m *OpenShiftManager) Process(
+	ctx context.Context,
+	nsname types.NamespacedName) Result {
+	// Do-nothing if not on OpenShift
+	if !IsOpenShiftCluster(ctx, m.client, m.cfg) {
```

Comment:
Stray thought, not required for this PR:  if we see that this gets callsed often enough it may be beneficial to cache the result of this call on the manager instance. We know that it "can't" change, so fetching it from memory ought to be way faster than reading from a k8s api call.

Reply:
I agree; however, we should keep in mind that this in-memory state is ternary: un-initialized, true or false.  At any given moment k8s-scheduler may decide to restarted the samba-operator as a different process on a different node.

Reply:
Agree 100%. I generally prefer to define a tristate with a custom type like `type MyTriState int` `const (StateUnknown = MyTriState(0); StateBad = ...; StateGood = ...)`.
Sometimes people like to use a `*bool`. Which is also valid, but I feel it is not as clear/"self documenting".